Verify openings

michaelsproul opened the pull request:
This is my WIP attempt at verifying openings. I've run out of time for this week, and got kind of stuck on actually computing a pairing (couldn't seem to find a `blst` binding for it). Will come back to it next week!

ralexstokes commented on src/commitment.rs:
```diff
@@ -4,10 +4,14 @@ use crate::polynomial;
+    pub input: point::Point,
     pub value: point::Point,
+    pub fs_in_g1: blst::blst_p1,
     pub proof: Proof,
```
theoretically, i'd like to hide the actual type for a `proof`, perhaps that is something like:

```rust
type Proof = blst::blst_p1;

pub struct Opening {
    pub value: point::Point,
    pub proof: Proof,
}
```

and leave `Proof` private?

michaelsproul replied:
I'm not a fan of `type` (and it doesn't hide anything). If we want to make `Proof` opaque I think we'd need a struct

```rust
pub struct Proof {
    // private field
    proof: blst::blst_p1,
}
```

For now I've just left it as a `blst::blst_p1`.

ralexstokes commented on src/commitment.rs:
```diff
@@ -4,10 +4,14 @@ use crate::polynomial;
+}
 
 pub struct Opening {
+    pub input: point::Point,
```
i'm not super attached to the following but my thinking is that the verifier will know which `input` they are querying for along w/ the commitment ahead of time.

so the protocol flow goes something like:

- ahead of time, prover and verifier have the shared setup.
- prover publishes the commitment
- verifier sends over a query `x`.
- prover responds with the minimal information required which is the evaluation `y = f(x)` and the proof.
- verifier can verify `f(x) = y` with this information.

which means, if we want to keep the `Opening` struct minimal, e.g. because it will be serialized to disk/network, then it only contains `y` and the proof.

michaelsproul replied:
Yep! Sounds good! I've updated the API accordingly

ralexstokes commented on src/commitment.rs:
```diff
@@ -56,6 +65,36 @@ pub fn create<'a>(polynomial: &'a polynomial::Polynomial, setup: &setup::Setup)
+
+            let mut neg_y_in_g1 = blst::blst_p1::default();
+            blst::blst_p1_mult(&mut neg_y_in_g1, g1, &y_scalar, constants::MODULUS_BIT_SIZE);
+            blst::blst_p1_cneg(&mut neg_y_in_g1, true); // FIXME(sproul): wtf does the bool do
```
i'm pretty sure that the flag is referring to one of these bits here: https://tools.ietf.org/html/draft-irtf-cfrg-pairing-friendly-curves-07#appendix-C.1

and in fact it may even be the `C_bit`... i would write a little test just to verify that this being `true` or `false` matches your intuition -- check out `point::negate` for an example (although in blst it seems that the high-order bit is used to indicate sign in `Fr`).

michaelsproul replied:
I think it might be a constant-time flag, but I'm not sure. Ben Edginton seems to use `true` consistently in `c-kzg`:

https://github.com/benjaminion/c-kzg/blob/66c50a49b1811d743fba195dca042d262674c2b4/src/bls12_381.c#L367

But I shall investigate further.

mratsim replied:
cneg is for conditional negation. If you want to always negate you need to pass true, if you pass false it will leave the input untouched.

michaelsproul replied:
awesome, thanks @mratsim!

ralexstokes reviewed (commented):
awesome start! left some comments

michaelsproul commented:
I found the functions to compute the pairings and I think my impl is pretty close to the Py reference and c-kzg, although currently the test is failing.

michaelsproul commented:
Lots of progress!

* Added a library to wrap `blst` called `oblast` (област)
* Re-implemented verification of openings in terms of `oblast` (much simpler!)
* Wrote a test to confirm that the verification works for a very simple polynomial (`test_verify_opening_identity_polynomial`)
* Fixed a bug in the setup that was causing it to read uninitialized memory (https://github.com/ralexstokes/kzg/pull/2/commits/829a3ca692683c2f1775ed3d770d6ca0f283d573)

TODO:

* Ripple oblast through the rest of the codebase (assuming you like the idea). We'll need to wrap some more operations, particularly for `Fr` (should be a matter of porting across the stuff from `point.rs`).
* Work out why the verification of the proof generated by `open_at` is failing. I think it _might_ be the proof generation? I haven't looked at that yet.

michaelsproul commented:
I think this is ready for review now. The tests are all passing after I changed the test secrets to be `< r` (i.e. part of the field `Fr`). I regenerated two new seeds, and generated tests with `py_ecc` for all the same polynomials.

ralexstokes commented on oblast/src/lib.rs:
```diff
@@ -0,0 +1,223 @@
+        point
+    }
+
+    pub fn from_raw(element: blst_fr) -> Self {
```
ah this is nice -- gives the caller the ability to provide something greater than a `u64` for the `Fr` initialization...

we will also probably want a `{to,from}_bytes` method but we can add this down the line

ralexstokes commented on oblast/src/lib.rs:
```diff
@@ -0,0 +1,223 @@
+
+            pub fn compress(&self) -> Vec<u8> {
+                let mut compressed_point = vec![0; $compressed_bytes];
+                let compress = paste! { blst::[<blst_ $blst_name _compress>] };
+                unsafe {
+                    compress(compressed_point.as_mut_ptr(), &self.point);
+                }
```
any overhead here to defining the fn to a binding? rather than just assembling the name in place? there may be some limitations on `paste` here i don't know...

michaelsproul replied:
I imagine the compiler optimises it, but there's no reason `paste!` couldn't be inside the `unsafe` block except that it breaks `rustfmt` so I was avoiding it for that reason.

There are probably a few `#[inline]` sprinklings and things that we can do to make sure we're squeezing every bit of performance out too

ralexstokes commented on src/setup.rs:
```diff
@@ -11,6 +11,9 @@ pub struct Setup {
     let modulus = constants::get_modulus();
     let s = BigUint::from_bytes_be(secret);
+
+    assert!(s < constants::get_modulus(), "secret must be less than r");
```
nice catch!

ralexstokes reviewed (approved):
awesome work!
